src/parser/ion_1_0/text/mod.rs

```diff
@@ -0,0 +1,2614 @@
+///
+/// Most values are delimited by whitespace. Some are not, particularly in s-expressions. Since
+/// delimiting rules within s-expressions there are even more combinations. This method encapsulates
+/// value delimiting and returns one or two values, depending on if the next value was necessary
+/// for delimiting the first.
+///
+/// It would be an improvement to peek at the next character and just execute based on that,
+/// especially if we are dealing with any kind of token stream. That approach is more error prone,
+/// however, as we have to build the list of acceptable delimiting characters by hand.
+fn take_delimited_value<'a, F, D>(
+    value_parser: F,
+    delimiter_parser: D,
+    delimiter: Option<char>,
+    table: Table,
+) -> impl Fn(&'a str) -> IonResult<&'a str, (ion::Data, Option<ion::Value>)>
+where
+    F: Fn(&'a str) -> IonResult<&'a str, ion::Data>,
+    D: Fn(&'a str) -> IonResult<&'a str, ion::Data>,
+{
+    move |i: &str| {
+        // First check if we should continue by trying to parse a value.
+        let (i, data) = value_parser(i)?;
+
+        // Drop out if the next character is the optional delimiter
+        if let Some(delimiter) = delimiter {
+            if i.chars().next().map(|c| c == delimiter).unwrap_or(false) {
+                return Ok((i, (data, None)));
+            }
+        };
+
+        // Find the delimiter, which may be a quoted symbol for example
+        let (i, next) = cut(alt((
+            value(None, peek(many1(ws))),
+            map(
+                pair(
+                    take_quoted_annotation(table.clone()),
+                    take_value_parts(table.clone()),
+                ),
+                |(head_annotation, (mut rest_annotation, value))| {
+                    // It is mandatory to maintain the order of annotations applied to an object.
+                    rest_annotation.insert(0, Some(head_annotation));
+                    Some(ion::Value {
+                        value,
+                        annotations: Some(rest_annotation),
+                    })
+                },
+            ),
+            map(&delimiter_parser, |value| {
+                Some(ion::Value {
+                    value,
+                    annotations: None,
+                })
+            }),
+            // Addition (not from encoding):
+            // Inspired by 'good/timestamp/timestampWithTerminatingEof.ion'.
+            // A parsed value was terminated by the end of the file.
+            value(None, peek(eof)),
+        )))(i)?;
+
+        Ok((i, (data, next)))
+    }
+}
+
+/// value
+///     : annotation* entity
+///     ;
+fn take_value(table: Table) -> impl Fn(&str) -> IonResult<&str, ion::Value> {
+    move |i: &str| {
+        map(take_value_parts(table.clone()), |(annotations, value)| {
+            let annotations = if annotations.is_empty() {
+                None
+            } else {
+                Some(annotations)
+            };
```

Comment:
I think I prefer the latter, `Option<Vec<_>>` feels kind of redundant. I can make that change if it sounds good. Do we gain anything from `None` over `vec![]`?

Reply:
It's something of a vestigial holdover from the binary encoding. A binary encoded value either is or is not wrapped in one or more annotations. I found it convenient to distinguish between "having" annotations and "not having" annotations, since "having annotations but 0 of them" is a thing that could crop in the the binary encoding and should be rejected as invalid.

But I think `vec![]` makes for a better API experience, especially given that the text encoding doesn't seem to have that strange distinction between "having" zero annotations and not having annotations. Feel free to make the change.

http://amzn.github.io/ion-docs/docs/binary.html#annotations

Reply:
Alright, cleaned this up a bit. I added a `From<Data>` impl for `Value` a while back and used that in a number of places to simplify this, let me know if that's alright.